File: ride-service/src/main/java/com/example/ride_service/listener/RideConsumer.java
package com.example.ride_service.listener;

import com.example.ride_service.model.enums.EventType;
import com.example.ride_service.model.enums.TopicType;
import com.example.ride_service.model.event.DriverAssignedEvent;
import com.example.ride_service.service.RideService;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.messaging.handler.annotation.Header;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class RideConsumer {
    private final ObjectMapper objectMapper;
    private final RideService rideService;

    @KafkaListener(topics = TopicType.RIDE_LIFECYCLE_TOPIC)
    public void listen(String payload, @Header("eventType") String eventTypeString) {
        try {
            EventType eventType = EventType.fromEventName(eventTypeString);

            if (eventType == EventType.ASSIGNED_DRIVER) {
                log.info("Received {} event", eventType.getEventName());
                var assignDriverEvent = objectMapper.readValue(payload, DriverAssignedEvent.class);
                rideService.acceptRide(assignDriverEvent);
                log.info("Ride with id: {} accepted", assignDriverEvent.rideId());
            }

        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Unknown event type: " + eventTypeString);

        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }
}

File: driver-service/src/main/kotlin/com/example/driver_service/listener/RideConsumer.kt
package com.example.driver_service.listener

import com.example.driver_service.model.enums.CancelInitiator
import com.example.driver_service.model.enums.EventType
import com.example.driver_service.model.enums.WorkStatus
import com.example.driver_service.model.event.NoDriversEvent
import com.example.driver_service.model.event.RideCreateEvent
import com.example.driver_service.model.view.toAssignedDriverEvent
import com.example.driver_service.service.DriverMatchingService
import com.example.driver_service.service.DriverService
import com.example.driver_service.service.OutboxEventService
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Value
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.messaging.handler.annotation.Header
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class RideConsumer(
    private val objectMapper: ObjectMapper,
    private val driverMatchingService: DriverMatchingService,
    private val outboxEventService: OutboxEventService,
    private val driverService: DriverService,
    @param:Value("\${app.Kafka.ride-topic}") private val rideTopic: String
) {
    companion object {
        private val log = KotlinLogging.logger {}
    }

    @Transactional
    @KafkaListener(topics = ["\${app.Kafka.ride-topic}"])
    fun listen(
        payload: String,
        @Header("eventType") type: String
    ) {
        try {
            log.info { "Processing event: $type" }
            when (type) {
                EventType.RIDE_CREATED.eventName -> {
                    val event = objectMapper.readValue<RideCreateEvent>(payload)

                    val driver = driverMatchingService.findBestDriver(event.startPoint, event.seats)

                    if (driver != null) {
                        outboxEventService.saveEvent(
                            driver.toAssignedDriverEvent(event.rideId),
                            EventType.ASSIGNED_DRIVER,
                            rideTopic
                        )

                        driverService.setWorkStatus(driver.id, WorkStatus.BUSY)

                        log.info { "available driver with id: ${driver.id} is assigned to ride with id: ${event.rideId}" }
                    } else {
                        val noDriversEvent = NoDriversEvent(
                            event.rideId,
                            "available drivers not found",
                            CancelInitiator.SYSTEM
                        )
                        outboxEventService.saveEvent(
                            noDriversEvent,
                            EventType.NO_AVAILABLE_DRIVERS,
                            rideTopic
                        )
                        log.info { "available drivers not found!" }
                    }
                }
            }
        } catch (e: Exception) {
            log.error(e) { "Error while processing kafka message: ${e.message}" }
        }
    }
}